### PR TITLE
Replace allure-xunit's custom attributes with runner reporter (fixes #344)

### LINKED ISSUE #344: Get rid of allure-xunit's [AllureXunit] and [AllureXunitTheory] attributes
Currently, in order to use allure-xunit one have to:
1. Install the package
2. Create `allureConfig.json`
3. Replace all [Fact] and [Theory] attributes with [AllureXunit] and [AllureXunitTheory].

The goal of this issue is to get rid of the 3rd step.

## Motivation
The end goal is to make allure-xunit usable out-of-box. A user should be able to generate the report by just installing the package (optionally, adding some configuration) and running the tests.

### Negative implications of using [AllureXunit] and [AllureXunitTheory]
The need to replace attributes has two major negative implications:

1. It's a time-consuming operation if a user has a fairly large number of tests.
2. The [AllureXunit] and [AllureXunitTheory] attributes cannot be composed with other attributes derived from [Fact] or [Theory]. If a user decides to utilize such attributes from a 3rd party library, there will be no way to use allure-xunit at all.

## Example

1. Current usage:

    ```csharp
    using Allure.Xunit.Attributes;
    using Xunit;

    public class TestClass
    {
        [AllureXunit]
        public void TestMethod() { }

        [AllureXunitTheory]
        [InlineData(1, "a")]
        [InlineData(2, "b")]
        public void ParameterizedTestMethod(int a, string b) { }
    }
    ```

2. Future usage:

    ```csharp
    using Xunit;

    public class TestClass
    {
        [Fact]
        public void TestMethod(){}

        [Theory]
        [InlineData(1, "a")]
        [InlineData(2, "b")]
        public void ParameterizedTestMethod(int a, string b) { }
    }
    ```

The same report should be created for both usages:

## Backward compatibility

Allure-xunit must still recognize the [AllureXunit] and [AllureXunitTheory] attributes correctly and collect test results from the corresponding test methods.

### PR DESCRIPTION
**Pull request by delatrie**
This is the first piece of work to improve allure-xunit out-of-box experience. The main goal here is to drop the `[AllureXunit]` and `[AllureXunitTheory]` attributes in favor of the built-in ones: `[Fact]` and `[Theory]`.

### Context

Basically, to generate a report we need to listen for xunit's execution events and extract data provided by xunit in event messages. There are a couple of ways to achieve this.

#### Previous implementation

Custom attributes (`[AllureXunit]` and `[AllureXunitTheory]`) were used to discover tests via the `AllureXunitDiscover` creating an instance of `AllureXunitTestCase` for each and injecting the event listener (namely, `AllureMessageBus`) to handle the events.

The disadvantage is that in order to use allure-xunit one has to modify their test code first replacing built-in xunit attributes with custom ones.

#### New implementation

We drop the currently implemented extension mechanism in favor of a custom runner reporter (`AllureRunnerReporter`) that may be used by xunit to directly create an event handler (in our case, an instance of `AllureMessageSink`) allowing us to bypass the need of custom attributes.

The `[AllureXunit]` and `[AllureXunitTheory]` attributes are marked as obsoleted.

### Usage

Just install the package and run the tests as usual.

In some cases, allure reporter may not run automatically. In that case, enable the reporter explicitly:

  - For VS runner (in Visual Studio, or via the dotnet CLI) use `RunConfiguration.ReporterSwitch=allure` runsettings option.
  - For console runner use `-noautoreporters -allure` CLI arguments (make sure Allure.XUnit.dll and xunit.console.exe are in the same directory).
  - For MSBuild runner speficy the `NoAutoReporters="true" Reporter="allure"` properties in your `xunit` task definition (make sure Allure.XUnit.dll and xunit.runner.msbuild.net452.dll are in the same directory).

### Additional changes

The PR also adds support for static test methods:

```csharp
using Xunit;

public class MyTestClass
{
    [Fact]
    public static void StaticMethod()
    {
        Assert.Equal(3, 1 + 2);
    }
}

```

Previously, such a method caused a test run to be aborted.

#### One reporter per test run

Xunit allows only one active runner reporter. Currently it's impossible to produce allure output files AND, say, send the output to TeamCity at the same test run.

Runners are quite composable though. It should not be that hard to wrap 3rd party's event listener in our own. Support for that will be added in near future.

#### Arguments of some theories might potentially be unreported

Xunit doesn't provide test arguments to event listeners if one the following conditions is true:

  1. A user explicitly disabled theory data pre-enumeration during discovery phase (either on per-test basis or globally).
  2. Xunit can't serialize one of a test's argument.

In that case xunit creates one ITestCase instance with null `TestMethodArguments` property, and multiple ITest instances, one per each argument set. There is no public API allowing to get these arguments.

The PR workarounds this problem by attaching a hook to the XunitTestRunner's constructor at runtime using [Harmony][harmony]. Harmony has its own [limitations][harmony-edge-cases], the most concerning of which is the inability of Harmony to patch an inlined method. If the constructor of XunitTestRunner is inlined, the patch doesn't work, and arguments of a test that falls into one of the categories above are not reported by allure-xunit. The warning is issued in such a case advising to switch to debug mode.

#### Checklist
- [x] [Sign Allure CLA][cla]
- [ ] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure2
[harmony]: https://github.com/pardeike/Harmony
[harmony-edge-cases]: https://harmony.pardeike.net/articles/patching-edgecases.html
[harmony-2.3]: https://github.com/pardeike/Harmony/releases/tag/v2.3-prerelease.2

### CODE REVIEW

**Comment by neparij** [comment]
Add README notes.
Create issues from Known issues section

**Review by neparij — CHANGES_REQUESTED** [review]

File: Allure.XUnit/Allure.XUnit.csproj [hunk 1]
```diff
@@ -22,7 +22,9 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <PackageReference Include="Lib.Harmony" Version="2.3.0-prerelease.2" />
```
[the remark this answers is not in the document]
Comment (delatrie) [reply Allure.XUnit/Allure.XUnit.csproj:25]: Primarily because of net7.0 support by Harmony (see [here](https://github.com/pardeike/Harmony/releases/tag/v2.3-prerelease.2)). API is the same as in GA version (they only have switched to MonoMod.Core internally). I will create an issue to not to forget to update Harmony as soon as its 2.3 GA is released.

File: Allure.XUnit/AllureXunitHelper.cs [hunk 2]
```diff
@@ -31,142 +31,17 @@ static AllureXunitHelper()
             Environment.SetEnvironmentVariable(allureConfigEnvVariable, allureConfigPath);
         }
 
-        public static void StartTestContainer(ITestCaseStarting testCaseStarting)
```
[the remark this answers is not in the document]
Comment (delatrie) [reply Allure.XUnit/AllureXunitHelper.cs:34]: I've moved the code related to allure model manipulation back AllureXunitHelper to simplify AllureMessageSink a little bit. Public methods are refactored and marked as [Obsolete].

**Comment by delatrie** [comment]
> Add README notes.
> Create issues from Known issues section

Done!
See #368 and #369 and the updated [readme].

[readme]: https://github.com/allure-framework/allure-csharp/blob/de3222098e42c5da2e4ad6e4630bab26d90ff427/Allure.XUnit/README.md

**Review by neparij — APPROVED** [review]
